Base app corrections

matliw opened the pull request.

Commit:
Added working /users/ endpoint with GET and POST methods

Commit:
Added working endpoints with methods and users

Commit:
Recovery -

Commit:
Changing urlspatterns, adding new serializers and views

Commit:
Changed url patterns and views

Commit:
Added requirements.txt

Commit:
Refined models and serializers, added swagger-UI

Commit:
Changed name and description for swagger-ui API

Commit:
Create working API model, no parameter searches

Commit:
Fixed url paths and methods

Commit:
Added drf_spectacular to requirements.txt

Commit:
Corrected models/serializers/views/urls
Added migrations
Removed schema.yml

Commit:
Added authorization for memebers

ikonstantynov commented on quickbudget/migrations/0001_initial.py:
```diff
@@ -0,0 +1,54 @@
+# Generated by Django 4.1.7 on 2023-04-06 19:07
```
migrations are corrupted. please regenerate them.
the error:
```
It is impossible to change a nullable field 'budget_id' on expense to non-nullable without providing a default. This is because the database needs something to populate existing rows.
```

ikonstantynov commented on quickbudget/urls.py:
```diff
@@ -0,0 +1,17 @@
+    path("api/budgets/<uuid:budget_id>/expenses/", AddExpenses.as_view(), name="list_add_budget_expenses"),
+    path("api/budgets/<uuid:budget_id>/expenses/<uuid:expense_id>", ExpenseDetail.as_view(), name="get_update_remove_expense_details"),
+    path("api/categories/", ListCategoryList.as_view(), name="category_list"),
+    path('api/docs/', SpectacularAPIView.as_view(), name='schema'),  # yaml schema is generated
+    path('api/docs/swagger-ui/', SpectacularSwaggerView.as_view(url_name='schema'), name='swagger-ui'),  # swagger-ui endpoint for API visualization
```
api/schema - for schema
api/docs - for ui

Commit:
Added permissions

Commit:
Regenerated initial migration

Commit:
Merged from grant_access

Commit:
Added query_set method to views

Commit:
Added user post/delete endpoint
Added pagination
Added customn permission for non-auth/auth in specific methods
Added receiver with last interaction
Added JWT token auth

Commit:
Added user post/delete endpoint
Added pagination
Added customn permission for non-auth/auth in specific methods
Added receiver with last interaction
Added JWT token auth

Commit:
Added filter expenses by category
Added order expenses by created_timestamp

Commit:
Corrected paths in urls

Commit:
Added Search expense endpoint
Added ordering for expenses

Commit:
feat: save User.id in 'Budget.creator' column when creating a new budget object

Commit:
feat: only creator can add and remove budget members

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+
+
+    def put(self, request, budget_id, format=None):
+        budget = Budget.objects.get(id=budget_id)
```
you should check if the budget exists

matliw replied:
added check using:
```
def get_object(self, budget_id):
        try:
            return Budget.objects.get(id=budget_id)
        except Budget.DoesNotExist:
            raise Http404
```

ikonstantynov commented on requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Django~=4.1.7
+djangorestframework~=3.14.0
+drf_spectacular~=0.26.1
```
missing `djangorestframework-simplejwt`

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+                return Response(serializer.data)
+
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        return Response(serializer.errors, status=status.HTTP_403_BAD_REQUEST)
```
no such status

matliw replied:
oops

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+                return Response(serializer.data)
+
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        return Response(serializer.errors, status=status.HTTP_403_BAD_REQUEST)
```
no such status

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+
+
+class ListAddBudgets(generics.ListCreateAPIView):
+    queryset = Budget.objects.all()
```
you don't need it since you overwrite the logic in `get_queryset`

matliw replied:
removed

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+
+        return Response(serializer.data, status=status.HTTP_400_BAD_REQUEST)
+
+    def get_queryset(self):
+        return Budget.objects.filter(members=self.request.user)
```
should be above all the methods for visibility

matliw replied:
moved

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+
+
+class BudgetDetail(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Budget.objects.all()
```
copying it from your previous PR:
```
you need to be careful when usiing `queryset = Budget.objects.all()`. in some cases, you'd need to overwrite `get_queryset` method and filter data.
```

try to debug what data is in there, and what data you expect.

matliw replied:
It was displaying all the data and with get_queryset we can filter for data that can be viewed by currently logged in user.

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+        serializer = MemberAddSerializer(budget, data=request.data)
+        self.check_object_permissions(request, budget)
+
+        if self.request.user.id == budget.creator_id:
```
better to update to the following:
```
if self.request.user.id != budget.creator_id:
    return Response(serializer.errors, status=status.HTTP_403_FORBIDDEN)
```

in this case you don't need to put the main code/logic inside the `if` block.

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+    def put(self, request, budget_id, format=None):
+        budget = Budget.objects.get(id=budget_id)
+        serializer = MemberAddSerializer(budget, data=request.data)
+        self.check_object_permissions(request, budget)
```
what's the point of checking the THIS perms here?

matliw replied:
without it the APIView is not checking permissions for these methods

matliw replied:
removed

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+    def delete(self, request, budget_id, format=None):
+        budget = Budget.objects.get(id=budget_id)
+        serializer = RemoveBudgetMembersSerializer(budget, data=request.data)
+        self.check_object_permissions(request, budget)
```
what's the point of checking the THIS perms here?

ikonstantynov commented on quickbudget/api/permissions.py:
```diff
@@ -0,0 +1,47 @@
+        if request.user.is_superuser:
+            return True
+
+        if request.user in obj.members.all():
```
I'd add checking if a user is a budget owner/creator.
+
I'd rename the class. A superuser or a budget creator could not be a budget member. It's more comprehensive to call it: "CanAccessBudget".
+ 
currently, a budget creator can delete all members (including themself), so they cannot access the budget after that.

ikonstantynov commented on quickbudget/api/serializers.py:
```diff
@@ -0,0 +1,90 @@
+        fields = ["id", "name", "description"]
+
+
+class \
+        MemberAddSerializer(serializers.ModelSerializer):
```
please don't do that. there's no reason to move text on a new line

matliw replied:
it's my ide doing it. I've installed black linter

ikonstantynov commented on quickbudget/receivers.py:
```diff
@@ -0,0 +1,11 @@
+@receiver(post_save, sender=Expense)
+def interaction_with_budget(sender, instance, **kwargs):
+    Budget.objects.get(id=instance.budget_id.id).save(
+        update_fields=["last_interaction"]
```
you update this value but I don't see any ordering when accessing a budget

ikonstantynov commented on quickbudget/urls.py:
```diff
@@ -0,0 +1,27 @@
+    path("api/budgets/<uuid:budget_id>/expenses/<uuid:expense_id>/", ExpenseDetail.as_view(), name="get-update-remove-expense-details"),
+    path("api/categories/", ListCategoryList.as_view(), name="category-list"),
+    path("api/schema/", SpectacularAPIView.as_view(), name="schema"),  # yaml schema is generated
+    path("api/docs/swagger-ui/", SpectacularSwaggerView.as_view(url_name="schema"), name="swagger-ui"),  # swagger-ui endpoint for API visualization
```
without `swagger-ui`

ikonstantynov commented on quickbudget/urls.py:
```diff
@@ -0,0 +1,27 @@
+    path("api/schema/", SpectacularAPIView.as_view(), name="schema"),  # yaml schema is generated
+    path("api/docs/swagger-ui/", SpectacularSwaggerView.as_view(url_name="schema"), name="swagger-ui"),  # swagger-ui endpoint for API visualization
+    path("api/users/", QuickbudgetUsers.as_view(), name="application-user-add-remove-members"),
+    path("api/search/", SearchBudgetExpenses.as_view(), name="search-for-expenses")
```
better to be explicit when name endpoints
`api/search/expenses` or smth like that

ikonstantynov commented on quickbudget/models.py:
```diff
@@ -0,0 +1,53 @@
+    created_timestamp = models.DateTimeField(auto_now=True)
+    members = models.ManyToManyField(settings.AUTH_USER_MODEL)
+    last_interaction = models.DateTimeField(auto_now=True)
+    creator = models.ForeignKey(User, null=True, related_name="budget_owner", on_delete=models.CASCADE)
```
`created_by`

ikonstantynov commented on quickbudget/models.py:
```diff
@@ -0,0 +1,53 @@
+    budget_id = models.ForeignKey(Budget, on_delete=models.CASCADE, related_name="expenses")
+    category = models.ForeignKey(Category, null=True, on_delete=models.SET_NULL)
+    created_timestamp = models.DateTimeField(auto_now=True)
+    author = models.ForeignKey(User, null=True, on_delete=models.CASCADE, related_name="expense_owner")
```
`created_by`

ikonstantynov commented on quickbudget/models.py:
```diff
@@ -0,0 +1,53 @@
+    name = models.CharField(max_length=120)
+    total = models.DecimalField(max_digits=8, decimal_places=2, validators=[MinValueValidator(Decimal('0.01'))])
+    description = models.CharField(max_length=200, null=True, blank=True)
+    budget_id = models.ForeignKey(Budget, on_delete=models.CASCADE, related_name="expenses")
+    category = models.ForeignKey(Category, null=True, on_delete=models.SET_NULL)
```
as far as I remember, we discussed that the naming of FK must be the same

ikonstantynov commented on quickbudget/models.py:
```diff
@@ -0,0 +1,53 @@
+from core import settings
+
+
+class User(AbstractUser):
+
```
nice to have: instead of incremental user id, you should use UUID

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+        # Set the creator field to the current authenticated user
+        serializer.validated_data['creator'] = self.request.user
+
+        if serializer.is_valid():
+            # Create the budget object
```
```
if not serializer.is_valid():
    return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+    serializer_class = ExpenseSerializer
+    permission_classes = [AllBudgetExpenseMembersOnly]
+    filter_backends = (filters.OrderingFilter,)
+    ordering_fields = ["name", "total", "created_timestamp"]
```
```I believe all expenses should be ordered by `created_timestamp` DESC```

sorry, I guess this comment should relate to the default ordering.

ikonstantynov commented on quickbudget/views.py:
```diff
@@ -0,0 +1,162 @@
+
+    def get_queryset(self):
+        budget = self.kwargs["budget_id"]
+        queryset = Expense.objects.filter(budget_id=budget).order_by(
+            "created_timestamp")
+        category = self.request.query_params.get("category")
+        if category is not None:
```
you can replace this code with Ordering settings
https://www.django-rest-framework.org/api-guide/filtering/#specifying-a-default-ordering

ikonstantynov commented on quickbudget/api/serializers.py:
```diff
@@ -0,0 +1,90 @@
+class UserSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = User
+        fields = ["id", "username", "password"]
```
you need to create another serializer for responses. you don't need to return the password (even if it's hashed).

ikonstantynov replied:
also, look at the `.to_representation` method for the BaseSerializer class.

ikonstantynov commented on quickbudget/api/permissions.py:
```diff
@@ -0,0 +1,47 @@
+        if request.user.is_superuser:
+            return True
+
+        if request.user in obj.shopping_list.members.all():
```
`shopping_list`

Commit:
fix: creator cannot remove himself from the member list

Commit:
feat: changed swagger url

Commit:
fix: removed .is_valid() nesting

Commit:
fix: permissions for budget access

Commit:
fix:
1. changed fields creator/author to created_by
2. removed redundant model instancing
3. added pylint and black to reqs
4. serializers.py - formatting
5. urls.py extended url path

Commit:
feat: set default ordering to "last_interaction"

Commit:
fix: changed field names in Budget model

Commit:
fix: removed password from being serialized in to_representation

Commit:
fix: added migration

Commit:
docker, debug tool, postgres

Commit:
fix: db username

ikonstantynov commented on requirements-dev.txt:
```diff
@@ -0,0 +1,3 @@
+black==22.6.0
+flake8==4.0.1
+isort==5.10.1
```
add this 
```
django-debug-toolbar==4.0.0

-r requirements.txt
```

ikonstantynov commented on requirements.txt:
```diff
@@ -0,0 +1,12 @@
+psycopg2-binary==2.9.3
+black==22.6.0
+flake8==4.0.1
+isort==5.10.1
```
remove from the file

ikonstantynov commented on requirements.txt:
```diff
@@ -0,0 +1,12 @@
+djangorestframework==3.14.0
+drf_spectacular~=0.26.1
+djangorestframework-simplejwt~=5.2.2
+django-debug-toolbar==4.0.0
```
remove from the file

ikonstantynov replied:
django-debug-toolbar must not be in prod settings. it's just for local dev

ikonstantynov commented on dockerfile:
```diff
@@ -0,0 +1,72 @@
+# pull official base image
```
must starts with the capital letter: `Dockerfile`

Commit:
fix: serializer_class in BudgetMemebers view

Commit:
fix changed requirements, small fixes

matliw committed:
N+1 problem adjustments

matliw committed:
Messages added to 403 responses in /members endpoint

matliw committed:
fix: updating and deleting expenses
